File: java_converter/src/main/java/com/mdeditor/javaconverter/ReportWriter.java
package com.mdeditor.javaconverter;

import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.List;

final class ReportWriter {
  private ReportWriter() {
  }

  static void write(Path vault, ProjectModel project, AnalysisResult result) throws IOException {
    write(vault, project, new ConversionResult(
        result.sources().size(),
        result.sources().size(),
        result.localDependencyCount(),
        result.unresolvedCount(),
        java.time.Instant.EPOCH,
        java.time.Instant.EPOCH,
        java.time.Duration.ZERO,
        result.diagnostics(),
        result.unresolvedSymbols()
    ));
  }

  static void write(Path vault, ProjectModel project, ConversionResult result) throws IOException {
    List<String> lines = new ArrayList<>();
    lines.add("# Java Converter Report");
    lines.add("");
    lines.add("- Root: " + project.root());
    lines.add("- Started: " + ConversionClock.formatInstant(result.startedAt()));
    lines.add("- Finished: " + ConversionClock.formatInstant(result.finishedAt()));
    lines.add("- Duration: " + ConversionClock.formatDuration(result.duration()));
    lines.add("- Source files: " + project.sourceFiles().size());
    lines.add("- Files analyzed: " + result.filesAnalyzed());
    lines.add("- Markdown files: " + result.markdownFilesWritten());
    lines.add("- Local dependencies: " + result.localDependencyCount());
    lines.add("- Unresolved symbols: " + result.unresolvedCount());
    lines.add("- Classpath entries: " + project.classpathEntries().size());
    lines.add("- Maven modules detected: " + project.mavenModules());
    lines.add("");

    appendSection(lines, "Warnings", project.warnings());
    appendSection(lines, "Unresolved Symbols", result.unresolvedSymbols());
    appendSection(lines, "Compiler Diagnostics", result.diagnostics());

    Files.writeString(vault.resolve("_java_converter_report.md"), String.join("\n", lines), StandardCharsets.UTF_8);
  }

  private static void appendSection(List<String> lines, String title, List<String> values) {
    lines.add("## " + title);
    lines.add("");
    if (values.isEmpty()) {
      lines.add("None.");
    } else {
      values.forEach(value -> lines.add("- " + value));
    }
    lines.add("");
  }
}

File: desktop-app/converters/code_converter/lib/report-writer.js
const fs = require("fs");
const path = require("path");
const { MD_EDITOR_DIR, REPORT_JSON_FILE, REPORT_MARKDOWN_FILE } = require("./constants");
const { toMarkdownPath } = require("./utils");

function writeMissingDependenciesReport(destinationRoot, report) {
  const metadataDir = path.join(destinationRoot, MD_EDITOR_DIR);
  fs.mkdirSync(metadataDir, { recursive: true });
  const jsonPath = path.join(metadataDir, REPORT_JSON_FILE);
  const markdownPath = path.join(metadataDir, REPORT_MARKDOWN_FILE);
  fs.writeFileSync(jsonPath, JSON.stringify(report, null, 2) + "\n", "utf8");

  const lines = [
    "# Built-in Converter Missing Dependencies Report",
    "",
    `- Root: ${report.root}`,
    `- Source root home: ${report.sourceRootPath}`,
    `- Started: ${report.startedAt}`,
    `- Finished: ${report.finishedAt}`,
    `- Duration: ${report.duration}`,
    `- Source files: ${report.counts.sourceFiles}`,
    `- Markdown files: ${report.counts.markdownFilesWritten}`,
    `- Local dependencies: ${report.counts.localDependencies}`,
    `- Unresolved dependencies: ${report.counts.unresolvedDependencies}`,
    `- Discovered external dependencies: ${report.counts.discoveredExternalDependencies}`,
    "",
    "## Warnings",
    "",
  ];
  if (report.warnings?.length) report.warnings.forEach((warning) => lines.push(`- ${warning}`));
  else lines.push("None.");
  lines.push("", "## Unresolved Dependencies", "");
  const unresolved = (report.sources || []).flatMap((source) => (
    (source.unresolvedDependencies || []).map((dependency) => ({ source, dependency }))
  ));
  if (!unresolved.length) {
    lines.push("None.");
  } else {
    unresolved.forEach(({ source, dependency }) => {
      lines.push(`- \`${dependency.symbol}\` (${dependency.language}, ${dependency.kind}) in ${source.sourceFile}`);
    });
  }
  fs.writeFileSync(markdownPath, lines.join("\n"), "utf8");
  return { jsonPath, markdownPath };
}

function createReport({ sourceRoot, sourceRootHome, destinationRoot, files, startedAt, finishedAt, sources, externalDependencies, warnings }) {
  const localDependencyCount = sources.reduce((total, source) => total + (source.localDependencies?.length || 0), 0);
  const unresolvedCount = sources.reduce((total, source) => total + (source.unresolvedDependencies?.length || 0), 0);
  return {
    schemaVersion: 1,
    generator: "builtin_converter",
    language: "mixed",
    root: sourceRoot,
    sourceRootPath: toMarkdownPath(sourceRootHome),
    vault: destinationRoot,
    startedAt,
    finishedAt,
    duration: `${Math.max(0, new Date(finishedAt) - new Date(startedAt))}ms`,
    counts: {
      sourceFiles: files.length,
      filesAnalyzed: files.length,
      markdownFilesWritten: sources.length,
      localDependencies: localDependencyCount,
      unresolvedDependencies: unresolvedCount,
      discoveredExternalDependencies: externalDependencies.length,
    },
    warnings,
    diagnostics: [],
    sources,
    externalDependencies,
  };
}

module.exports = {
  writeMissingDependenciesReport,
  createReport,
};
